node/core/dispute-coordinator/src/scraping/mod.rs

```diff
@@ -72,6 +69,60 @@ impl ScrapedUpdates {
 	}
 }
 
+/// A structure meant to facilitate chain reversions in the event of a dispute
+/// concluding against a candidate. Each candidate hash maps to a vector of block
+/// numbers and hashes for all blocks which included the candidate. The entries
+/// in each vector are ordered by decreasing parent block number to facilitate
+/// minimal cost pruning.
+pub struct InclusionsPerCandidate {
+	inclusions_inner: HashMap<CandidateHash, Vec<(BlockNumber, Hash)>>,
+}
+
+impl InclusionsPerCandidate {
+	pub fn new() -> Self {
+		Self { inclusions_inner: HashMap::new() }
+	}
+
+	// Insert parent block into vector for the candidate hash it is including,
+	// maintaining ordering by decreasing parent block number.
+	pub fn insert(
+		&mut self,
+		candidate_hash: CandidateHash,
+		block_number: BlockNumber,
+		block_hash: Hash,
+	) {
+		if let Some(blocks_including) = self.inclusions_inner.get_mut(&candidate_hash) {
+			for idx in 0..blocks_including.len() {
+				if blocks_including[idx].0 < block_number {
+					// Idx is between 0 and blocks_including.len(), therefore in bounds. QED
+					blocks_including.insert(idx, (block_number, block_hash));
```

Comment:
nit: I have another Rust trick for you 🤓: [`partition_point`](https://doc.rust-lang.org/1.66.1/core/primitive.slice.html#method.partition_point). (Note this assumes the slice is sorted, which this one is.)